### DISCUSSION TITLE
How to add "incorrect_spans" feature into pretrained spacy model

### DISCUSSION THREAD

**Discussion by Ashufet** [explosion/spaCy#9088] [Help: Model Advice] [answered]
Hi Team,

Currently, I am using the pre-trained spacy model "en_core_web_lg" for NER prediction. There are some words for which model is giving the wrong prediction like -

"TestData" being predicted as "Person".

I wanted to use the latest feature("incorrect_spans") of spacy but I am confused about how to do that. As per documentation,
I need to update config file of ner as mentioned below -

```
from spacy.pipeline.ner import DEFAULT_NER_MODEL
config = {
   "moves": None,
   "update_with_oracle_cut_size": 100,
   "model": DEFAULT_NER_MODEL,
   "incorrect_spans_key": "incorrect_spans",
}
nlp.add_pipe("ner", config=config)
```

But when I am doing it, I am getting an error as "ner is already added in pipe", which seems reasonable but then how to update ner config file.

Can anyone help here?

**Comment by svlandeg** [comment, marked as answer]
Hi,

I'm afraid this process is not as smooth yet as we would like (and we definitely want to address this in the future).

If I understand your use-case correctly, you want to fine-tune an existing NER model. That means you don't want to use `add_pipe`, because that will create a blank model from scratch. Instead, you want to create a new pipeline with your required components by creating a [config file](https://spacy.io/usage/training#config) that [sources ](https://spacy.io/usage/processing-pipelines#sourced-components) the relevant components:
```
[components.ner]
source = "en_core_web_lg"
component = "ner"
```
This will pull in the trained NER component from the other pipeline into your new pipeline.

Now, unfortunately the currently pretrained pipelines (e.g. `en_core_web_lg 3.1.0) have this setting by default:
```
incorrect_spans_key = null
```
and there is currently no way yet to change it. We will implement this in the future. For now, a kind-of-silly workaround is to either edit your `en_core_web_lg` package directly or make a copy and edit that. When you do, make the edit in two places:
- change the `incorrect_spans_key` in the `config.cfg` in the root directory
- make the same change in the `cfg` file in the `ner` subdirectory.

Once that's done, you can `source` that (edited) `ner` component as described, prepare your training data with incorrect examples in `doc.spans["incorrect_spans"]`, mix in normal correct examples to avoid [catastrophic forgetting](https://explosion.ai/blog/pseudo-rehearsal-catastrophic-forgetting), and run [`spacy train`](https://spacy.io/api/cli#train) to hopefully get a better, fine-tuned model!

**Reply by Ashufet** [reply]
Thanks for the update, surely will try the workaround way and update here soon.

**Reply by ywsada** [reply]
Having the exact same problem, is this functionality implemented now, 2 years later? Or do I still need to perform the workaround? Thanks in advance.

**Reply by svlandeg** [reply]
As far as I know we haven't actually gotten around to implementing a better workflow for this yet. Sorry - but it's a good reminder for our (ever growing ;-)) TODO list!

**Reply by ywsada** [reply]
I see, the workaround is fine for the time being, thanks for your response!
